src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+                    var documentSymbols = DocumentOutlineHelper.GetNestedDocumentSymbols(responseBody);
+                    LatestFetchedDocumentModel = DocumentOutlineHelper.GetDocumentSymbolModels(documentSymbols);
+
+                    LspSnapshot = activeTextView.TextSnapshot;
+                    LatestFetchedDocumentModelIsNotEmpty = LatestFetchedDocumentModel.Length > 0;
+
+                    StartModelUpdateTask();
+                }
+                else
+                {
+                    LatestFetchedDocumentModelIsNotEmpty = false;
+                    LatestFetchedDocumentModel = ImmutableArray<DocumentSymbolViewModel>.Empty;
+                }
+            }
+            else
+            {
+                LatestFetchedDocumentModelIsNotEmpty = false;
+                LatestFetchedDocumentModel = ImmutableArray<DocumentSymbolViewModel>.Empty;
+            }
+
+            string? GetFilePath(IWpfTextView textView)
+            {
+                ThreadingContext.ThrowIfNotOnUIThread();
+                if (textView.TextBuffer.Properties.TryGetProperty(typeof(IVsTextBuffer), out IVsTextBuffer bufferAdapter) &&
+                    bufferAdapter is IPersistFileFormat persistFileFormat &&
+                    ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out _)))
+                {
+                    return filePath;
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Processes the fetched document model and updates the UI.
+        /// </summary>
+        private async ValueTask UpdateModelAsync(CancellationToken cancellationToken)
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+            {
+                var updatedSymbolTreeSource = LatestFetchedDocumentModel;
+
+                // Switch to UI thread to obtain the search query.
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                var searchQuery = searchBox.Text;
+
+                // Switch to the threadpool to get the lastest text view and filter and sort the model.
+                await TaskScheduler.Default;
+                CodeWindow.GetLastActiveView(out var textView);
+                var activeTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+                if (activeTextView is not null)
+                {
+                    if (!string.IsNullOrWhiteSpace(searchQuery))
+                        updatedSymbolTreeSource = DocumentOutlineHelper.Search(updatedSymbolTreeSource, searchQuery);
+
+                    updatedSymbolTreeSource = DocumentOutlineHelper.Sort(updatedSymbolTreeSource, SortOption, LspSnapshot, activeTextView.TextSnapshot);
+
+                    await UpdateSymbolTreeSourceAsync(updatedSymbolTreeSource, cancellationToken).ConfigureAwait(false);
+                    StartHightlightNodeTask();
```

Comment:
Since we were already on the UI thread to get everything else, we didn't want to just figure out the node to highlight while we were there?

Reply:
Not really since we also call `StartHightlightNodeTask()` whenever the user moves the caret around so I don't think we should duplicate that code.